src/Stratis.SmartContracts.IntegrationTests/ContractInternalTransferTests.cs

```diff
@@ -1,31 +1,231 @@
+            uint256 currentHash = this.node1.GetLastBlock().GetHash();
+
+            // Send amount to contract, which will send to contract address
+            string[] parameters = new string[] { string.Format("{0}#{1}", (int)SmartContractCarrierDataType.Address, receiveResponse.NewContractAddress) };
+            BuildCallContractTransactionResponse response = this.node1.SendCallContractTransaction(
+                nameof(BasicTransfer.SendToAddress),
+                preResponse.NewContractAddress,
+                amount,
+                parameters);
+            this.node2.WaitMempoolCount(1);
+            this.node2.MineBlocks(1);
+
+            Block lastBlock = this.node1.GetLastBlock();
+
+            // Blocks progressed
+            Assert.NotEqual(currentHash, lastBlock.GetHash());
+
+            // Contract doesn't maintain any balance
+            Assert.Equal((ulong)0, this.node1.GetContractBalance(preResponse.NewContractAddress));
+
+            // Receiver contract now has balance
+            Assert.Equal((ulong) new Money((int) amount, MoneyUnit.BTC), this.node1.GetContractBalance(receiveResponse.NewContractAddress));
+
+            // Block contains a condensing transaction
+            Assert.Equal(3, lastBlock.Transactions.Count);
+            Transaction condensingTransaction = lastBlock.Transactions[2];
+            Assert.Single(condensingTransaction.Outputs); // Entire balance was forwarded
+            byte[] toBytes = condensingTransaction.Outputs[0].ScriptPubKey.ToBytes();
+            Assert.Equal((byte)ScOpcodeType.OP_INTERNALCONTRACTTRANSFER, toBytes[0]);
+            uint160 toAddress = new uint160(toBytes.Skip(1).ToArray());
+            Assert.Equal(receiveResponse.NewContractAddress, toAddress.ToAddress(this.mockChain.Network).Value);
+            Assert.Equal(new Money((long)amount, MoneyUnit.BTC), condensingTransaction.Outputs[0].Value);
+            Money fee = lastBlock.Transactions[0].Outputs[0].Value - new Money(50, MoneyUnit.BTC);
+
+            // Amount in wallet is reduced by amount sent and fee.
+            Assert.Equal(senderBalanceBefore - this.node1.WalletSpendableBalance, fee + new Money((long)amount, MoneyUnit.BTC));
+
+            // Receipt is correct
+            ReceiptResponse receipt = this.node1.GetReceipt(response.TransactionId.ToString());
+            Assert.Equal(lastBlock.GetHash().ToString(), receipt.BlockHash);
+            Assert.Equal(response.TransactionId.ToString(), receipt.TransactionHash);
+            Assert.Empty(receipt.Logs); // TODO: Could add logs to this test
+            Assert.True(receipt.Success);
+            Assert.True(receipt.GasUsed > GasPriceList.BaseCost);
+            Assert.Null(receipt.NewContractAddress);
+            Assert.Equal(this.node1.MinerAddress.Address, receipt.From);
+            Assert.Null(receipt.Error);
+            Assert.Equal(preResponse.NewContractAddress, receipt.To);
+        }
```

Comment:
Do you think it would make sense to additionally test if the receive handler was invoked?

Reply:
Good call - I'll log a receipt and save something to storage?

Reply:
Yep, sounds good